File: scripts/cheatsheet.py
"""Printable tiered cheat sheet -- the offline fallback.

If the laptop dies, the wifi drops, or the draft app eats the browser tab, this
is what you draft from. Open the HTML and print to PDF.
"""

from __future__ import annotations

import argparse
import html
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import polars as pl

from src.config import DATA_PROCESSED, LEAGUE, SEASON
from src.draft.replacement import add_vor, replacement_levels
from src.draft.sim_draft import snake_picks
from src.draft.tiers import add_tiers
from src.project.consensus import project
from src.project.kdst import project_kdst

POS_COLORS = {"QB": "#7c3aed", "RB": "#059669", "WR": "#2563eb",
              "TE": "#d97706", "DST": "#475569"}


def _tier_blocks(board: pl.DataFrame, pos: str, max_players: int) -> str:
    sub = board.filter(pl.col("pos") == pos).sort("proj_points", descending=True).head(max_players)
    out = []
    for tier, group in sub.group_by("tier", maintain_order=True):
        rows = "".join(
            f"<tr><td>{html.escape(r['name'])}</td><td class='t'>{r['tm']}</td>"
            f"<td class='n'>{r['adp']:.0f}</td><td class='n'>{r['proj_points']:.0f}</td>"
            f"<td class='n b'>{r['vor']:.0f}</td><td class='n'>{r['bye'] or '-'}</td></tr>"
            for r in group.sort("adp").iter_rows(named=True)
        )
        out.append(
            f"<div class='tier'><div class='tierhead'>Tier {tier[0]}"
            f"<span>{group.height} player{'s' if group.height > 1 else ''}</span></div>"
            f"<table><tr><th>Player</th><th>Tm</th><th>ADP</th><th>Proj</th>"
            f"<th>VOR</th><th>Bye</th></tr>{rows}</table></div>"
        )
    return "".join(out)


def build(season: int, slot: int | None, out_path: Path) -> Path:
    proj = project(season)
    try:
        kdst = project_kdst(season)
        if kdst.height:
            proj = pl.concat([proj, kdst], how="diagonal_relaxed")
    except Exception:
        pass
    board = add_tiers(add_vor(proj))
    levels = replacement_levels(board)

    picks = snake_picks(slot, LEAGUE["teams"], LEAGUE["rounds"]) if slot else []
    picks_html = (
        "<div class='picks'><b>Your picks (slot " + str(slot) + "):</b> "
        + " &middot; ".join(str(p) for p in picks) + "</div>"
        if picks else ""
    )

    cols = "".join(
        f"<section><h2 style='border-color:{POS_COLORS[p]}'>"
        f"<span style='color:{POS_COLORS[p]}'>{p}</span></h2>"
        f"<div class='repl'>replacement: {levels.get(p, 0):.0f} pts</div>"
        f"{_tier_blocks(board, p, n)}</section>"
        # Defences earn a column now: year-over-year persistence is 0.276, more
        # than twice the kicker's 0.109. Kickers stay off the sheet deliberately.
        for p, n in [("RB", 42), ("WR", 50), ("TE", 20), ("QB", 20), ("DST", 16)]
    )

    doc = f"""<!doctype html><meta charset="utf-8"><title>Cheat Sheet {season}</title>
<style>
 @page{{size:A4 landscape;margin:10mm}}
 body{{font:10px/1.3 ui-sans-serif,system-ui,sans-serif;color:#111;margin:0}}
 h1{{font-size:15px;margin:0 0 2px}}
 .sub{{color:#666;font-size:10px;margin-bottom:6px}}
 .picks{{background:#f4f6f8;border:1px solid #dde3ea;padding:5px 8px;
        border-radius:5px;margin-bottom:8px;font-size:10px}}
 .grid{{display:grid;grid-template-columns:repeat(4,1fr);gap:10px}}
 section{{break-inside:avoid}}
 h2{{font-size:12px;margin:0 0 2px;padding-bottom:2px;border-bottom:2px solid}}
 .repl{{color:#888;font-size:9px;margin-bottom:5px}}
 .tier{{margin-bottom:7px;break-inside:avoid}}
 .tierhead{{display:flex;justify-content:space-between;background:#eef1f5;
   padding:2px 5px;font-weight:700;font-size:9px;border-radius:3px}}
 .tierhead span{{color:#7a8593;font-weight:400}}
 table{{width:100%;border-collapse:collapse;margin-top:2px}}
 th{{font-size:8px;color:#8a94a0;text-align:right;font-weight:600;padding:1px 3px}}
 th:first-child,td:first-child{{text-align:left}}
 td{{padding:1px 3px;border-bottom:1px solid #f0f2f5}}
 .n{{text-align:right;font-variant-numeric:tabular-nums}}
 .t{{color:#8a94a0;font-size:9px}}
 .b{{font-weight:700}}
</style>
<h1>{season} Draft Cheat Sheet &mdash; {LEAGUE['teams']}-team full PPR</h1>
<div class="sub">Tiers are groups the historical data cannot separate: take any
player in a tier over reaching for one in the tier below. VOR is value over this
league's replacement level. Never take the last player in a tier if the next tier
is still deep.</div>
{picks_html}
<div class="grid">{cols}</div>"""

    out_path.write_text(doc, encoding="utf-8")
    return out_path


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--season", type=int, default=SEASON)
    ap.add_argument("--slot", type=int, default=None)
    args = ap.parse_args()

    DATA_PROCESSED.mkdir(parents=True, exist_ok=True)
    path = build(args.season, args.slot, DATA_PROCESSED / "cheatsheet.html")
    print(f"Wrote {path}")
    print("Open it in a browser and print to PDF (A4 landscape).")


if __name__ == "__main__":
    main()

File: v2/lib/cheatsheet.ts
/**
 * A self-contained cheatsheet, for when the network is not there.
 *
 * Hosting this tool trades away the best property of the thing it replaces:
 * `src/dashboard/app.py` runs on a laptop with no network, no install and
 * nothing to go wrong on the morning of the draft. This buys part of that back.
 * One HTML file, no scripts, no fetches, no fonts -- open it or print it, and
 * it works if Vercel is down, if the database is down, or if the draft venue's
 * wifi is the usual disaster.
 *
 * It is deliberately not a copy of the live board. It carries what you cannot
 * recompute in your head: the ordering, the tier breaks, and where replacement
 * sits.
 */

import { untieredPositions } from "./tiers";
import type { LeagueConfig, Player } from "./types";

const DEFAULT_LIMIT = 150;

function escapeHtml(value: string): string {
  return value
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

function snakeRounds(slot: number, teams: number, rounds: number): number[] {
  const picks: number[] = [];
  for (let r = 0; r < rounds; r++) {
    picks.push(r % 2 === 0 ? r * teams + slot : r * teams + (teams - slot + 1));
  }
  return picks;
}

export interface CheatsheetInput {
  leagueName: string;
  league: LeagueConfig;
  players: Player[];
  slot: number;
  adpAsOf: string | null;
  /** Anchor-format warnings, so the printed copy carries them too. */
  warnings?: string[];
  limit?: number;
}

export function renderCheatsheet(input: CheatsheetInput): string {
  const { league, slot } = input;
  const limit = input.limit ?? DEFAULT_LIMIT;

  const board = [...input.players].sort((a, b) => (b.vor ?? 0) - (a.vor ?? 0));
  const top = board.slice(0, limit);

  const replacement = new Map<string, number>();
  for (const p of input.players) {
    if (p.replacement !== undefined) replacement.set(p.pos, p.replacement);
  }

  const picks = snakeRounds(slot, league.teams, league.rounds);

  // The heavy rule marks the last player of a tier -- the one case worth
  // reaching for.
  //
  // Stated as "the next printed row is a different tier" this is very nearly
  // right, and was: tier-mates share a projection by definition, so at a fixed
  // position they share a VOR and sort into one run. It is right by luck of the
  // ordering rather than by construction, though, and it misses the last row on
  // the page and any tier still running past the cut. Counting a tier's members
  // and marking the row that exhausts it says the same thing on purpose.
  const tierTotal = new Map<string, number>();
  for (const p of input.players) {
    const key = `${p.pos}|${p.tier}`;
    tierTotal.set(key, (tierTotal.get(key) ?? 0) + 1);
  }
  const tierSeen = new Map<string, number>();
  // Positions the fit could not pool at all -- every kicker his own tier. The
  // rule would mark every one of their rows, which is not a tier break, it is
  // just the list continuing.
  const untiered = untieredPositions(input.players);

  const rows = top
    .map((p, i) => {
      const key = `${p.pos}|${p.tier}`;
      const seen = (tierSeen.get(key) ?? 0) + 1;
      tierSeen.set(key, seen);
      const tiered = !untiered.has(p.pos);
      // Counted against the whole board, not the printed slice: a tier that
      // runs on past the cut has not been exhausted by the last row shown.
      const tierBreak = tiered && seen === tierTotal.get(key);
      return `<tr class="${tierBreak ? "tier-end" : ""}">
      <td class="num">${i + 1}</td>
      <td>${escapeHtml(p.name)}</td>
      <td class="pos ${p.pos}">${p.pos}</td>
      <td>${escapeHtml(p.tm ?? "")}</td>
      <td class="num">${p.adp.toFixed(1)}</td>
      <td class="num">${p.proj_points.toFixed(0)}</td>
      <td class="num dim">${(p.sd ?? 0).toFixed(0)}</td>
      <td class="num">${(p.vor ?? 0).toFixed(0)}</td>
      <td class="num">${tiered ? `${p.pos}${p.tier}` : "&mdash;"}</td>
      <td class="num">${p.bye ?? "-"}</td>
      <td class="num dim">${(p.playoff_lift ?? 0) > 0 ? "+" : ""}${(p.playoff_lift ?? 0).toFixed(1)}</td>
    </tr>`;
    })
    .join("\n");

  const replacementRows = [...replacement.entries()]
    .sort((a, b) => (a[0] < b[0] ? -1 : 1))
    .map(([pos, level]) => `<tr><td>${pos}</td><td class="num">${level.toFixed(1)}</td></tr>`)
    .join("\n");

  const pickList = picks
    .map((p, i) => `<li><span class="rd">R${i + 1}</span> pick ${p}</li>`)
    .join("\n");

  const warnings = (input.warnings ?? [])
    .map((w) => `<li>${escapeHtml(w)}</li>`)
    .join("\n");

  return `<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>${escapeHtml(input.leagueName)} - draft cheatsheet</title>
<style>
  :root { color-scheme: light; }
  * { box-sizing: border-box; }
  body {
    margin: 0; padding: 24px;
    font: 12px/1.45 -apple-system, BlinkMacSystemFont, "Segoe UI", Helvetica, Arial, sans-serif;
    color: #16181d; background: #fff;
  }
  h1 { font-size: 18px; margin: 0 0 2px; }
  h2 { font-size: 12px; text-transform: uppercase; letter-spacing: .08em;
       color: #6b7280; margin: 0 0 8px; }
  .meta { color: #6b7280; font-size: 11px; margin-bottom: 18px; }
  .layout { display: grid; grid-template-columns: 1fr 210px; gap: 28px; align-items: start; }
  table { border-collapse: collapse; width: 100%; }
  th, td { text-align: left; padding: 3px 6px; border-bottom: 1px solid #eef0f3; }
  th { font-size: 10px; text-transform: uppercase; letter-spacing: .06em; color: #6b7280;
       border-bottom: 1px solid #d6dae1; }
  td.num, th.num { text-align: right; font-variant-numeric: tabular-nums; }
  tr.tier-end td { border-bottom: 2px solid #b9c0cb; }
  td.dim { color: #8a919c; }
  .pos { font-weight: 600; font-size: 10px; }
  .QB { color: #b3452f; } .RB { color: #1f7a4d; } .WR { color: #2454a6; }
  .TE { color: #8a5a1a; } .K  { color: #6b7280; } .DST { color: #55507a; }
  aside section { margin-bottom: 22px; }
  ul { margin: 0; padding-left: 16px; }
  li { margin-bottom: 2px; }
  .rd { display: inline-block; width: 26px; color: #6b7280; }
  .warn { border-left: 3px solid #d08b2c; padding-left: 10px; color: #6b5426; }
  @media print {
    body { padding: 0; font-size: 10px; }
    .layout { gap: 18px; }
    tr { page-break-inside: avoid; }
  }
</style>
</head>
<body>
  <h1>${escapeHtml(input.leagueName)}</h1>
  <div class="meta">
    ${league.teams}-team &middot; slot ${slot} &middot; ${league.rounds} rounds
    &middot; ADP as of ${escapeHtml(input.adpAsOf ?? "unknown")}
    &middot; ranked by value over replacement
  </div>

  <div class="layout">
    <main>
      <h2>Top ${top.length} by VOR</h2>
      <table>
        <thead>
          <tr>
            <th class="num">#</th><th>Player</th><th>Pos</th><th>Tm</th>
            <th class="num">ADP</th><th class="num">Proj</th><th class="num">&plusmn;</th>
            <th class="num">VOR</th><th class="num">Tier</th><th class="num">Bye</th>
            <th class="num">Playoff</th>
          </tr>
        </thead>
        <tbody>
${rows}
        </tbody>
      </table>
    </main>

    <aside>
      <section>
        <h2>Your picks</h2>
        <ul>${pickList}</ul>
      </section>
      <section>
        <h2>Replacement level</h2>
        <table><tbody>${replacementRows}</tbody></table>
      </section>
      ${
        warnings
          ? `<section class="warn"><h2>Read first</h2><ul>${warnings}</ul></section>`
          : ""
      }
      <section>
        <h2>Reading it</h2>
        <ul>
          <li>A heavy rule under a player means he is the last of his tier. Being last in a tier is the only good reason to reach.</li>
          <li>Proj is read off a curve of draft rank against what players at that rank actually scored, so a tier shares one number. &plusmn; is the measured spread around it — wide enough that the projection is the middle of a range, not a forecast.</li>
          <li>Inside a tier the projection is tied, so there is nothing to choose on but Bye and Playoff — how much softer weeks 15-17 are for his team than the rest of its season. Neither is priced into the ranking.</li>
          <li>VOR is against replacement, not against the field.</li>
          <li>K and DST last. They are near-random year to year.</li>
        </ul>
      </section>
    </aside>
  </div>
</body>
</html>`;
}
